docs/tutorials/data_sources/index.rst

```diff
@@ -3,6 +3,74 @@
 Data Sources
 ============
 
+A `Grain` data source is responsible for retrieving individual records. Records
+could be in a file/storage system or generated on the fly. Data sources need to
+implement the following protocol:
```

Comment:
we have two paths: RandomAccessDataSource -- it is supported by both, DataLoader and Dataset APIs and another path for sources that do not support random access -- the implement grain.IterDataset -- example of these are Parquet and tfrecord sources. They are only supported in Dataset API -- let's mention them as well. Th main difference is that sequential sources do not support global shuffle and instead use a hierarchical shuffle described here:

https://github.com/google/grain/blob/3edd5bccd67ec97b8b11215c82bf551b988c261c/grain/_src/python/dataset/sources/parquet_dataset_test.py#L89